convolutions session (#38)
nfidd/nfidd: sessions/04-using-delay-distributions-to-model-the-data-generating-process-of-an-epidemic/index.qmd

## Delay distributions and convolutions

In the last session we simulated _individual outcomes_ from a delay distribution, and then re-estimated the corresponding parameters.
However, sometimes we do not have data on these individual-level outcomes, either because they are not recorded or because they cannot be shared, for example due to privacy concerns.
At the population level, individual level delays translate into _convolutions_.
If we have a time series of infections $I_t$ ($t=1, 2, 3, \ldots, t_\mathrm{max}$), where $t$ denotes the day on which the infections occur, and observable outcomes occur with a delay given by a delay distribution $p_i$ ($i=0, 1, 2, \dots, p_\mathrm{max}$), where $i$ is the number of days after infection that the observation happens, then the number of observable outcomes $C_t$ on day $t$ is given by

$$
C_t = \sum_{i=0}^{i=p_\mathrm{max}} I_{t-i} p_i
$$

In words, to get the number of observable outcomes on day $t$ is given by the sum of infections on all previous days multiplied by the probability that those infections are observed on day $t$.

> For example, the observable outcomes $C_t$ could be the number of symptom onsets on day $t$ and $p_i$ is the incubation period.

Reviewer: why might we want to use a convolution?

Author: do you have a suggestion for a wording?

Reviewer: Will promote to an issue

We can use exactly the same data as in the previous session but this time aggregate it into a time series:

```{r aggregate}
data(infection_times)
## assume a start date, e.g. 10 May of this year
start_date <- ymd("2024-05-10")
df <- infection_times |>
  mutate(
    incubation_period = rgamma(n(), shape = 5, rate = 1),
    infection_date = floor_date(start_date + infection_time, unit = "day"),

>     onset_date = floor_date(

Reviewer: I think in an earlier session we could introduce a function to do this times -> date conversion and then use it? That way people can understand the operation as part of the censoring session and not have it as visual clutter as we go on?

Author: I have moved away from dates now, to be introduced in #54 instead.

      start_date + infection_time + incubation_period,
      unit = "day"
    ),
  )
## infection time series
inf_time_series <- df |>
  count(infection_date)
head(inf_time_series)
```

The resulting time series does contain any days with no infections.

> We can make our calculations easier by adding them:

Reviewer: ```suggestion
We can make our calculations easier (_as otherwise we would have to try and detect these in any models that used this data which could be complicated_)by adding them:
```

```{r missing_to_zero}

> all_dates <- expand(

Reviewer: This isn't really part of our teaching journey so I would wrap in a helper function or use one from elsewhere.

  df, infection_date = seq(min(infection_date), max(infection_date), by = "day")
)
inf_time_series <- all_dates |>
  full_join(inf_time_series, by = join_by(infection_date)) |>
  replace_na(list(n = 0))
head(inf_time_series)
```

Now we can convolve the time series with a delay distribution of outcomes to get a time series of outcomes as suggested above.
However, if we want to use the gamma distribution with shape 5 and rate 1 as before we face a problem.
The gamma distribution is a _continuous_ distribution but days are _discrete_ entities.

> One solution to this problem is to first _discretise_ the gamma distribution.

Reviewer: why are we introducing this like they haven't see the censoring session?

Author: phrasing is gone now.

> To do so we integrate it over 2-day windows, which will give a reasonable approximation to using the continuous distribution with continuous infection times and continuous outcomes.

Reviewer: Is this what it approximates or does it approximate double interval censoring?

Author: The phrasing (and method) is gone now.

> ```{r discretise_2_day_window}

Reviewer: We don't actually need to use this approximation if just doing it in R. We can either do it via simulation (as via the helper in `Epinowcast` or numerical integration (as in `EpiAware`). Whilst a bit more complex code wise (and stochastic in the first case) this might be nice as it avoids introducing a new (and not really correct) concept.

Reviewer: I think its a judgement call though

Author: This is a good idea I think. I've implemented the random approach.

    ## get vector of infections (14-day window)
    first_index <- max(1, i - max_delay)
    ts_segment <- ts[seq(first_index, i)]
    ## take reverse of pmf and reverse if needed
    pmf <- rev(delay)[seq_len(i - first_index + 1)]
    ## convolve with delay distribution
    ret <- sum(ts_segment * pmf)
    return(round(ret))
  }, numeric(1))
  return(convolved)
}
```

::: {.callout-tip}
## Take 5 minutes
Try to understand the `convolve_with_delay()` function above.
Try it with a few different time series and delay distributions.
How would you create the time series of infections using the discretised gamma distribution created above and saved in `gamma_pmf`?
:::

::: {.callout-note collapse="true"]}
```{r applied_convolution}
onsets <- convolve_with_delay(inf_time_series$n, gamma_pmf)
```
:::

We can plot these symptom onsets
```{r convolution_plot}
combined <- inf_time_series |>
  rename(date = infection_date, infections = n) |>
  mutate(onsets = onsets)
ggplot(combined, aes(x = date, y = onsets)) +
  geom_bar(stat = "identity")
```

Do they look similar to the plot of symptom onsets in the previous session?

## Observation uncertainty

> Usually not all data are perfectly observed.
> The counting process itself generally introduces a natural form of error that we can, once again, model with a probability distribution.
> For modelling counts, a common choice is the Poisson distribution.

Author: ```suggestion
Usually not all data are perfectly observed.
Also, the convolution we applied is a _deterministic_ operation that brushes over the fact that individual delays are random.
Given that we are now dealing with count data a natural choice for modelling the variation these processes introduce is the Poisson distribution.
```

We can use this to generate uncertainty around our convolved data.

```{r uncertain}
onsets <- onsets |>
  mutate(observed = rpois(n(), onsets))
```

Does a plot of these observations look more like the data from the previous session than the convolution plotted above?

# Estimating a time series of infection

> We now invert the procedure of above, i.e. instead of convolving infects and adding uncertainty to simulate observed symptom onsets, we start with observed simulated onsets to estimate the number of infections, given an incubation period.

Reviewer: ```suggestion
We now invert the procedure of above, i.e. instead of convolving infections and adding uncertainty to simulate observed symptom onsets, we start with observed simulated onsets to estimate the number of infections, given an incubation period.
```

> We now invert the procedure of above, i.e. instead of convolving infects and adding uncertainty to simulate observed symptom onsets, we start with observed simulated onsets to estimate the number of infections, given an incubation period.

Reviewer: I would really rather we not describe any form of model as backwards process as it seems misleading. 

really we start with somee prior model of infections and work forward based on what we know about the observation process. I think this will make the various add ons to the model seem a lot more logical than backwards steps and also makes it clearer what the dangers are

Author: Have rephrased.

Mathematically this amounts to a _deconvolution_ of the time series, i.e. the reverse process to convolution.

```{r stan_deconvolve}
mod <- cmdstan_model(
  here(
    "sessions",
    paste0(
      "04-using-delay-distributions-to-model-the-data-generating-process-of-",
      "an-epidemic"
    ),
    "estimate_infections.stan"
  )
)
mod$print(line_numbers = TRUE)
```

::: {.callout-tip}
## Take 10 minutes
Familarise yourself with the model above.

> Unlike before there is now a `functions` block at the beginning of the model (lines 1-33), where we define two functions, `discretise_gamma()` (line 2) and `convolve_with_delay()` (line 18).

Reviewer: For simplicity here I thought we would use the `pmf` calculated above? I don't mind hugely either way but this approach does add some model complexity for people to read through and maybe distracts from the core point? Or do you think teaching people about uncertainty propagation is worth it?

Author: Good idea.

> These correspond to our earlier **R** function `discretise()` and the code we used to generate a time series of symptom onsets, respectively.

Reviewer: I think I would prefer to describe this as `double_interval_censor` or something explicitly?

Author: now called censored_delay_pmf
